### ISSUE TITLE
pxr plugin, exporting & reimporting dag nodes changes hierarchy

### ISSUE THREAD

**Issue by exavi** [Autodesk/maya-usd#221]
I was wondering if I'm missing something obvious or the default I/O settings are wrong.

I have tried exporting some test geo, say a polyCube and on import it adds an additional transform.

Exported:
-pCube1 (transform)
    -pCubeShape1 (mesh)

After Import:
-pCube1 (transform)
    -pCubeShape1 (transform)
        -pCubeShape1Shape (mesh)

This seemingly harmless hierarchy change means that as meshes names are modified all our ldev assignment tools break, making testing or even considering USD usage a bit of a nightmare.

I'm sure this is some setting that can be avoided but honestly haven't been able to find what could stop this from happening as documentation for the plugin seems scarce, is there a way to avoid this?

When importing this into houdini everything shows up as intended, exporting the USD as ascii has confirmed that the export "looks right" so it seems to be an issue on import, unfortunately there is only 1 option on import and changing it does not seem to do the trick, anything I might be missing?

The USDa looks like this:
```
#usda 1.0
(
    defaultPrim = "pCube1"
    metersPerUnit = 0.01
    upAxis = "Y"
)

def Xform "pCube1" (
    kind = "component"
)
{
    def Mesh "pCubeShape1"{}
}
```

Tried using the latest builds of repos USD and maya-usd as of 28 Jan 2020.

**Comment by BigRoy** [comment]
@exavi regarding `maya-usd` which branch are you using? Is it `master`? If so, maybe it could be worth trying `dev` or `refactoring_sandbox` to see if changes would have fixed this already.

I'll see if I can reproduce this with the latest `dev` branch later today. Looking at the `.usda` file content you posted that does seem alright, so it seems like an issue on Import indeed.

**Comment by exavi** [comment]
@BigRoy I (think) I tried it with `dev` since it's the one that checks out by default on this repo and the one that builds with latest USD as of Today.

The latest "release" v0.02 and `master` seem to error saying that `UsdShadeMaterial::Bind()` does not exist anymore. _(I've confirmed with an older checkout of USD it did so the error makes sense)_

**Comment by BigRoy** [comment]
Great. And to be entirely sure:

1. Which of the plug-ins are you using inside of `dev`? :) Animal Logic's, Pixar or the autodesk one?
2. Do you happen to have the import command (e.g. mel/python commands being run) available too?

**Comment by exavi** [comment]
Thanks for your quick replies!

1. Pixar plugin as per the subject of the "issue" ;) But I'll be testing it with the Autodesk one in a bit. (if that's something it does, as I haven't seen much info about Autodesks one!)
2. Running:
`file -import -type "pxrUsdImport"  -ignoreVersion -ra true -mergeNamespacesOnClash false -namespace "geomtest_cube" -options ";shadingMode=none;readAnimData=0;useAsAnimationCache=0;assemblyRep=Import;startTime=0;endTime=0;useCustomFrameRange=0"  -pr  -importFrameRate true  -importTimeRange "override" "/homes/javiere/Desktop/geomtest_cube.usda";`

I'm currently building USD19.11 and will be building the older versions of this repo to see if this still happens. USD takes its time to build.

**Comment by exavi** [comment]
Built USD19.11 and maya-usd release 0.02 and this was already an issue back in Oct/Nov 19.

The following usda on import creates an additional transform that wasn't part of the exported hierarchy: (it also does it for much more complex USDs)

```
#usda 1.0
(
    defaultPrim = "pCube1"
    metersPerUnit = 0.01
    upAxis = "Y"
)

def Xform "pCube1" (
    kind = "component"
)
{
    def Mesh "pCubeShape1"
    {
        uniform bool doubleSided = 1
        float3[] extent = [(-0.5, -0.5, -0.5), (0.5, 0.5, 0.5)]
        int[] faceVertexCounts = [4, 4, 4, 4, 4, 4]
        int[] faceVertexIndices = [0, 1, 3, 2, 2, 3, 5, 4, 4, 5, 7, 6, 6, 7, 1, 0, 1, 7, 5, 3, 6, 0, 2, 4]
        point3f[] points = [(-0.5, -0.5, 0.5), (0.5, -0.5, 0.5), (-0.5, 0.5, 0.5), (0.5, 0.5, 0.5), (-0.5, 0.5, -0.5), (0.5, 0.5, -0.5), (-0.5, -0.5, -0.5), (0.5, -0.5, -0.5)]
        color3f[] primvars:displayColor = [(0.13320851, 0.13320851, 0.13320851)] (
            customData = {
                dictionary Maya = {
                    bool generated = 1
                }
            }
        )
        float2[] primvars:st = [(0.33, 0), (0.66333336, 0), (0.66333336, 0.25), (0.33, 0.25), (0.66333336, 0.5), (0.33, 0.5), (0.66333336, 0.75), (0.33, 0.75), (0.66333336, 1), (0.33, 1), (1, 0), (1, 0.25), (0, 0), (0, 0.25)] (
            interpolation = "faceVarying"
        )
        int[] primvars:st:indices = [0, 1, 2, 3, 3, 2, 4, 5, 5, 4, 6, 7, 7, 6, 8, 9, 1, 10, 11, 2, 12, 0, 3, 13]
    }
}

```

**Comment by fowlertADSK** [comment]
Thanks for logging something.  I'll investigate this one.

**Comment by fowlertADSK** [comment]
Took a quick look…  I’m not sure if this is exactly “By Design” but I think it’s always been this way, and is a result of the -mergeTransformAndShape option being off.  The default behavior on export is to only export a Mesh and include any transform information in it.  If you disable that option then we end up exporting a separate XForm and Mesh.

That makes sense to me, but it looks like the importer does not have a matching option to decide if it should automatically create the parent transform when it sees a Mesh prim, or just to parent it to the previously defined transform if it exists.  It looks like it’s hardcoded to always generate the transform+mesh pair of nodes in Maya when it seems a Mesh prim.  So you export with that option off and get a Xform and Mesh prims.  Then on import we see the Xform and create a Maya transform node, but then see the Mesh and create a transform+mesh pair and end up with two transforms.

I don’t know anything about the history of the option and why it was originally made to work this way (and I might have missed a flag or something that controls this behavior), but can you confirm that you were exporting from Maya with the -mergeTransformAndShape flag off?  If so, can you enable it and try again?

**Comment by exavi** [comment]
That's sorts it out thanks! mergeTransformAndShape was indeed off by default, I didn't expect it to be an export option given that the usda "looked" fine with a separate transform so I didn't go back to look at all export options. 
Perhaps it would make sense to have either a matching setting on import or do a little poll to see if people would be interested in this default changing in the future? 
Cheers!

**Comment by BigRoy** [comment]
@exavi wouldn't this still cause issues where the name on import starts to differ due to the shape still being split up from the transform? It all seems a bit magical.

Similarly wouldn't this still give trouble loading e.g. a USD from Houdini and getting it back to them in the same manner. In the end what matters is that the `.usd` file is formatted correctly. We shouldn't need to break what we expect in the `.usd` file solely to have this importer fix how it misbehaves. Or am I misinterpreting what went wrong in the end?

Could you share two simple `.usda` files, one with the setting enabled and one disabled? Just to see what we are dealing with exactly.

**Comment by exavi** [comment]
@BigRoy apologies for the delay, I tried it just as I was leaving the office and just got back. 

When exporting with "merge shapes & transforms" checked the output looks like so:

def Mesh "pCube1" (kind = "component")
{}

whereas with it disabled it looks like this:

def Xform "pCube1" ( kind = "component")
{
    def Mesh "pCubeShape1"
    {}
}

I think that on the Maya side this should be ok *_as long as shapes have not been manually renamed_* since it will automatically create shapes and assume that shapes should have the transform names concatenated with "*Shape" at the end...

If shapes have been manually renamed, it of course breaks by making name assumptions, (just confirmed it) so at this point I'm not sure if I should reopen the issue or not, the whole thing seems a bit dodgy either way.

Ideally the change/fix would be on the import side, as the output looked right in the first place, the "merged" one... not so much.

**Comment by BigRoy** [comment]
> If shapes have been manually renamed, it of course breaks by making name assumptions, (just confirmed it) so at this point I'm not sure if I should reopen the issue or not, the whole thing seems a bit dodgy either way.

This could turn into a problem also when adding a deformer to a shape. In some scenarios then produces a ` *DeformedShape` name.

Again, technically Maya cannot represent the mesh shape as a transform so it'll have to separate out the mesh into its own shape separate from the transform. In a way it does make sense. It should just try and reproduce a stable export as much as possible.

But thanks for confirming the behaviors.

Be aware that when you'd reference the first file into an Xform Prim that it will not show a mesh.
```
def Mesh "pCube1" (kind = "component")
{}
```
That is correct USD behavior as I believe referencing would not change the type of the Referencing prim unless you'd Reference it into a non-typed Primitive so it has no local override to an Xform type. This has confused some of our artists at the time, but made sense as they got what exactly was going on.

Basically it would import all your mesh data, however since the type is still Xform it means that a viewport would not know it has data to display as geometry, so you'll see nothing.

For simplicity with "Merge Shape to Transform" I'd recommend to make sure you have a least a Top group for the data. It will avoid this confusion much more.
